src/simularium/FrameRecorder.ts

```diff
@@ -0,0 +1,144 @@
+import { ArrayBufferTarget, Muxer } from "mp4-muxer";
+import { DEFAULT_FRAME_RATE } from "../constants";
+
+/**
+ * Records frames to an MP4 file using the WebCodecs API.
+ * Resulting file is passed to the front end implementation
+ * for download/use, via the handleFile() callback.
+ *
+ *
+ * Note that the VideoCodecs API is unavailable in some browsers, including Firefox,
+ * as of 2/6/2024. Viewport will not call these methods on firefox.
+ *
+ */
+export class FrameRecorder {
+    private getCanvas: () => HTMLCanvasElement | null;
+    private handleFile: (videoBlob: Blob) => void;
+    private encoder: VideoEncoder | null;
+    private muxer?: Muxer<ArrayBufferTarget>;
+    public isRecording: boolean;
+    private frameIndex: number;
+    public supportedBrowser: boolean;
+    private frameRate: number;
+
+    constructor(
+        getCanvas: () => HTMLCanvasElement | null,
+        handleFile: (videoBlob: Blob) => void
+    ) {
+        this.getCanvas = getCanvas;
+        this.handleFile = handleFile;
+        this.encoder = null;
+        this.isRecording = false;
+        this.frameIndex = 0;
+        this.supportedBrowser = !/firefox|fxios/i.test(navigator.userAgent);
+        this.frameRate = DEFAULT_FRAME_RATE;
+    }
+
+    private async setup(): Promise<void> {
+        const canvas = this.getCanvas();
+        if (canvas) {
+            try {
+                // VideoEncoder sends chunks of frame data to the muxer.
+                // Previously made one encoder in the constructor but
+                // making a new one during setup() prevents a bug where
+                // frames returned blank (stale canvas reference?)
+                this.encoder = new VideoEncoder({
+                    output: (chunk, meta) => {
+                        if (this.isRecording && this.muxer) {
+                            this.muxer?.addVideoChunk(chunk, meta);
+                        }
+                    },
+                    error: (error) => {
+                        console.error("Encoder error:", error);
+                    },
+                });
+                const config: VideoEncoderConfig = {
+                    codec: "avc1.420028",
+                    width: canvas.width,
+                    height: canvas.height,
+                };
```

Comment:
May be a good idea to check for hardware acceleration and to define stuff like the framerate of the resulting video here.
Using the avc codec makes sense and should be broadly supported, but might be worth doing a check for official documentation on that.

https://github.com/allen-cell-animated/nucmorph-colorizer/blob/main/src/colorizer/recorders/Mp4VideoRecorder.ts#L43-L91

Reply:
Defining frame rate definitely seems sensible, but I feel like when I do I can see a subtle reduction in output quality, more blur during camera motions. It's subtle so it could be in my head, maybe I can show you tomorrow in the office? Can you think why that would be?

As for hardware acceleration, it sounds like a good thing, but I find the docs confusing. [These docs](https://www.w3.org/TR/webcodecs/#hardware-acceleration) say `Setting a value of [prefer-hardware] or [prefer-software] can significantly restrict what configurations are supported.`

I think the fall back for that failed config is just to prefer software? But if there is a risk of a configuration error is the hardware acceleration worth it? I find all my videos download really fast while leaving this unconfigured.